Some steps towards Python 3 support

mgedmin opened the pull request:
This makes 'tox -e py34' pass the compileall step successfully and
actually start running the tests (plus fixes some, but not all, of the
test failures).

I want to keep this PR small and as close to obviously-correct as
possible, to avoid the fate of #10838.

mgedmin committed:
Use 'except ... as' syntax

This syntax works on Python 2.6 through 3.x.  lib/ansible/module_utils
(and lib/ansible/modules) need to support Python 2.4, so I didn't touch
those.

mgedmin committed:
Don't mix tabs and spaces

It's not allowed in Python 3 and merely a bad idea in Python 2.

mgedmin committed:
Use 0oNNN octal syntax

This syntax is valid in Python 2.6+ and 3.x.

mgedmin committed:
Drop the L suffix on numerical constants

Python has had automatic int-to-long promotion for a long long time now.
Even Python 2.4 does that automatically.

Python 3 drops support for the L suffix altogether.

mgedmin committed:
Support print as a function

I neglected the __future__ import because

  print(one_thing)

works the same way even when print is a statement.

mgedmin committed:
Avoid types.NoneType

types.NoneType was removed in Python 3.

None is a singleton in Python, so 'x is None' is equivalent to
'isinstance(x, NoneType)'.

mgedmin committed:
Fix to_bytes(None) on Python 3

You cannot call bytes(obj) to get a simple representation of obj on
Python 3!  E.g. bytes(42) returns a byte string with 42 NUL characters
instead of b'42'.

bcoca commented on v1/ansible/utils/template.py:
```diff
@@ -344,7 +343,7 @@ def template_from_string(basedir, data, vars, fail_on_undefined=False):
                 if var_name in vars:
                     resolved_val = vars[var_name]
-                    if isinstance(resolved_val, (bool, Number, NoneType)):
+                    if resolved_val is None or isinstance(resolved_val, (bool, Number)):
```
remove this file from the PR, v1 is dead code and will not be updated nor released, it is there just for quick testing.

mgedmin replied:
reverted this bit and rebased.

abadger commented on lib/ansible/utils/unicode.py:
```diff
@@ -215,7 +215,7 @@ def to_bytes(obj, encoding='utf-8', errors='replace', nonstring=None):
         try:
-            simple = binary_type(obj)
+            simple = str(obj)
         except UnicodeError:
```
This one is wrong as using str() would return the text type on python3 and this function should only return byte type on both python versions.

mgedmin replied:
There's code below this bit (lines 229-230) that converts text to bytes, if that's the case:

``` python
         if isinstance(simple, text_type):
             simple = simple.encode(encoding, 'replace')
```

abadger commented:
One comment that needs changing but I think everything else is good.

abadger commented:
@mgedmin Oh -- If the compileall test passes with these changes you can go ahead and enable them as well in your PR.

mgedmin commented:
@abadger: Enable them how/where?  tox.ini runs compileall on py34 before running nosetests (and did that before my changes).  Is there another place where compileall should be enabled?

mgedmin commented:
I see merge conflicts.  I'll rebase and fix them.

mgedmin committed:
Use 'except ... as' syntax in contrib/ and test/ too

mgedmin committed:
Support print() function in test/

mgedmin committed:
Use 0oNNN octal syntax in test/

mgedmin committed:
Disable compileall test for contrib/ in tox -e py34

(This is supposed to be temporary, until the code under contrib/ gets
cleaned up from all the remaining print statements.)

mgedmin committed:
Fix assert statement syntax

`assert (condition, message)` gets parsed by Python as `assert
a_two_tuple`, and a 2-element tuple is never False.

Discovered by compileall on Python 3.4, which emits a SyntaxWarning for
this common mistake.

abadger commented:
Nice work!  if we can enable the compileall tests in travis next then we hopefully won't keep regressing how much of our code compiles with python3.

# Merged

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
